### PR TITLE
[fix] fix gradient accumulation when update_frequency != 1

### PR DESCRIPTION
**Pull request by ytsheng**
* fix gradient accumulation when update_frequency is not equal to 1
  * The issue is caused by "start_update" being called at an irregular time so the gradients are actually not being accumulated in between update_frequency. When optimizer's `zero_grad` is called, grad is manually set to zero and therefore the gradients are not being accumulated. This fix made it so that `zero_grad` is called in the correct interval so that grad are being accumulated in between update_frequency.   
* fix `combined_report` not taking into account the new losses causing metrics calculation to be incorrect.
* add test (including logging)
  * add test to make sure there is a final update even if the num batches from the current update < update frequency.
* address issue: https://github.com/facebookresearch/mmf/issues/626

### CODE REVIEW

**Comment by vedanuj** [comment]
Thanks for the cleaner implementation. Where was the bug actually?

**Comment by ytsheng** [comment]
The issue is caused by "start_update" being called at an irregular time so the gradients are actually not being accumulated in between update_frequency. When optimizer's `zero_grad` is called, grad is manually set to zero and therefore the gradients are not being accumulated. This fix made it so that `zero_grad` is called in the correct interval so that grad are being accumulated in between update_frequency.

**Comment by vedanuj** [comment]
Thanks for the fix. Good catch! Is it possible to start a run with update frequency =2 for the movie mcan model? Just to make sure we get the desired result with this fix and can close the issue as well.

File: mmf/trainers/core/training_loop.py [hunk 1]
```diff
@@ -56,37 +56,19 @@ def run_training_epoch(self) -> None:
...
+            should_start_update = True
             for idx, batch in enumerate(self.train_loader):
-
-                if (idx + 1) % self.training_config.update_frequency == 0:
+                if should_start_update:
                     combined_report = None
-                    num_batches_for_this_update = min(
-                        self.training_config.update_frequency, num_remaining_batches
-                    )
-
```
Comment (vedanuj) [inline mmf/trainers/core/training_loop.py:82]: Is the case for remaining batches < self.training_config.update_frequency handled?

Comment (ytsheng) [reply mmf/trainers/core/training_loop.py:82]: @vedanuj, very good point - I think here the default behavior should be not doing the update if the number of iterations in the current update has not reached update_frequency yet -- bc users of this framework probably will not expect the last batch to have a high variance in their loss values.

What I mean is this: If the update_frequency is 100, and the number of batches we currently have done is 210, this means that we have already done 2 updates (2 * 100). And since 10 < 100, we will not be doing another update. Because the loss accumulated from the 10 iterations could be of high variance, which is less desirable to training especially having it be the last update. 

Let me know if you think otherwise. :)

**Comment by ytsheng** [comment]
Sample run in progress...

File: mmf/trainers/core/training_loop.py (hunk 1 above)
Comment (ytsheng) [reply mmf/trainers/core/training_loop.py:82]: I looked into pytorch lightning's [implementation](https://github.com/PyTorchLightning/pytorch-lightning/blob/e1c152b13c60f064a312f97ad0e326319bfda047/pytorch_lightning/trainer/training_loop.py#L853). It seems like they do perform a final update. Will update this PR.

**Comment by ytsheng** [comment]
@vedanuj, updated! :D Thanks for taking a look. Added a test too!

**Comment by ytsheng** [comment]
@vedanuj, at iteration 232000/236000, I am getting validation accuracy of 0.9078.

File: mmf/common/report.py [hunk 2]
```diff
@@ -84,3 +84,14 @@ def accumulate_tensor_fields(self, report, field_list):
                 continue
             if isinstance(self[key], torch.Tensor):
                 self[key] = torch.cat((self[key], report[key]), dim=0)
+
+    def accumulate_loss(self, report):
+        for key, value in report.losses.items():
+            if key not in self.losses.keys():
+                warnings.warn(
+                    f"{key} not found in report. Metrics calculation "
+                    + "might not work as expected."
```
Comment (vedanuj) [inline mmf/common/report.py:93]: Modify the warning message.

File: mmf/models/base_model.py [hunk 3]
```diff
@@ -169,7 +169,6 @@ def __call__(self, sample_list, *args, **kwargs):
         sample_list = to_device(sample_list, model_device)
 
         model_output = super().__call__(sample_list, *args, **kwargs)
-
```
Comment (apsdehal) [inline mmf/models/base_model.py:172]: Needed?

File: mmf/trainers/core/training_loop.py [hunk 4]
```diff
@@ -95,16 +95,22 @@ def run_training_epoch(self) -> None:
                     combined_report.accumulate_tensor_fields(
                         report, self.metrics.required_params
                     )
+                    # loss needs to be accumulated correctly for metric calculation
+                    combined_report.accumulate_loss(report)
```
Comment (apsdehal) [inline mmf/trainers/core/training_loop.py:99]: It would make sense to create a new `accumulate` function which does both as both are needed always. This will make sure both are never decoupled again.

File: tests/test_utils.py [hunk 5]
```diff
@@ -125,6 +125,19 @@ def forward(self, prepared_batch):
         return model_output
 
 
+class SimpleModelLogits(torch.nn.Module):
```
Comment (apsdehal) [inline tests/test_utils.py:128]: Inherit the SimpleModel and add stuff on top of it? (MSELoss doesn't even need to be assigned to module, you can directly use functional: https://pytorch.org/docs/stable/nn.functional.html#mse-loss)

**Review by apsdehal — APPROVED** [review]
Thanks for fixing this. Some minor comments and should be good to land.

**Review by vedanuj — APPROVED** [review]
Looks good. Please address Aman's comments before landing.
